### PR TITLE
Early errors for RegExp literals

### LINKED ISSUE #918: Invalid regexp property escape should throw an __early__ SyntaxError.
As verified in [Test262](https://github.com/tc39/test262/pull/1034#issuecomment-301541395), I'd like to confirm a mal formed property escape in a regexp should throw an Early Error.

I can't write a patch for this right now, but I want to keep track of this issue.

From @anba comments:

> ES 5.1 had:
> > An implementation must treat any instance of the following kinds of errors as an early error:
> > - [...]
> > - Errors in regular expression literals that are not implementation-defined syntax extensions.
>
> Somehow this was dropped in ES6 which leads to at least one contradiction, in https://tc39.github.io/ecma262/#sec-pattern, we've got:
> > [...] The algorithms in 21.2.2 are designed so that compiling a pattern may throw a SyntaxError exception; on the other hand, once the pattern is successfully compiled, applying the resulting internal procedure to find a match in a String cannot throw an exception (except for any host-defined exceptions that can occur anywhere such as out-of-memory).
>
> But a regular expression like `/a|(?![\d-\d])/u` may or may not evaluate the right-hand side disjunction (which contains an invalid range expression), it depends on the input string. 
> 
> So it should be listed as an early error in the spec, but no one spend time on correcting this issue...

### PR DESCRIPTION
**Pull request by anba**
Fixes #918

### CODE REVIEW

**Comment by littledan** [comment]
This patch is great--it seems to address this issue as well: https://github.com/tc39/proposal-regexp-unicode-property-escapes/issues/29

File: spec.html [hunk 1]
```diff
@@ -29130,6 +29130,266 @@
             It is a Syntax Error if _NcapturingParens_ &ge; 2<sup>32</sup>-1.
           </li>
         </ul>
+        <emu-grammar>QuantifierPrefix :: `{` DecimalDigits `,` DecimalDigits `}`</emu-grammar>
+        <ul>
+          <li>
+            It is a Syntax Error if the MV of the first |DecimalDigits| is larger than the MV of the second |DecimalDigits|.
+          </li>
+        </ul>
```
Comment (littledan) [inline spec.html]: From eshost, seems like four engines do have this early error.

**Comment by mathiasbynens** [comment]
I’d upvote this a thousand times if I could. Abstracting away `IsCharacterClass` etc. makes so much sense.

File: spec.html [hunk 2]
```diff
@@ -29130,6 +29130,266 @@
...
+        <emu-grammar>NonemptyClassRangesNoDash :: ClassAtomNoDash `-` ClassAtom ClassRanges</emu-grammar>
+        <ul>
+          <li>
+            It is a Syntax Error if IsCharacterClass of |ClassAtomNoDash| is *true* or IsCharacterClass of |ClassAtom| is *true*.
+          </li>
+          <li>
+            It is a Syntax Error if IsCharacterClass of |ClassAtomNoDash| is *false* and IsCharacterClass of |ClassAtom| is *false* and the CharacterValue of |ClassAtomNoDash| is larger than the CharacterValue of |ClassAtom|.
+          </li>
```
Comment (littledan) [inline spec.html]: Syntax error for everyone.

File: spec.html [hunk 3]
```diff
@@ -38970,16 +39070,67 @@
...
 
+      <emu-annex id="sec-patterns-static-semantics-early-errors-annexb">
+        <h1>Static Semantics: Early Errors</h1>
+        <p>The semantics of <emu-xref href="#sec-patterns-static-semantics-early-errors"></emu-xref> is extended as follows:</p>
+        <emu-grammar>ExtendedAtom :: InvalidBracedQuantifier</emu-grammar>
+        <ul>
+          <li>
+            It is a Syntax Error if any source text matches this rule.
+          </li>
+        </ul>
```
Comment (littledan) [inline spec.html]: Error in all tested engines.

File: spec.html [hunk 4]
```diff
@@ -38970,16 +39070,67 @@
...
+        <emu-grammar>NonemptyClassRanges :: ClassAtom `-` ClassAtom ClassRanges</emu-grammar>
+        <ul>
+          <li>
+            It is a Syntax Error if IsCharacterClass of the first |ClassAtom| is *true* or IsCharacterClass of the second |ClassAtom| is *true* <ins>and this production has a [U] parameter</ins>.
+          </li>
```
Comment (littledan) [inline spec.html]: This is what seems to be implemented in V8 and SpiderMonkey; the versions of JSC and ChakraCore I'm testing don't seem to throw the errors in non-Unicode mode.

File: spec.html [hunk 5]
```diff
@@ -38970,16 +39070,67 @@
...
+        </ul>
+        <emu-grammar>NonemptyClassRangesNoDash :: ClassAtomNoDash `-` ClassAtom ClassRanges</emu-grammar>
+        <ul>
+          <li>
+            It is a Syntax Error if IsCharacterClass of |ClassAtomNoDash| is *true* or IsCharacterClass of |ClassAtom| is *true* <ins>and this production has a [U] parameter</ins>.
+          </li>
```
Comment (littledan) [inline spec.html]: This is what seems to be implemented in V8 and SpiderMonkey; the versions of JSC and ChakraCore I'm testing don't seem to throw the errors in non-Unicode mode.

File: spec.html (hunk 4 above)
Comment (bterlson) [reply spec.html]: Probably don't want to commit these \<ins\> tags.

**Comment by bterlson** [comment]
Seems good. I just want to do my own littledan exercise of going through the various implementations and the early errors prescribed here.

Should this be a needs-consensus PR?

**Comment by bakkot** [comment]
@littledan, I don't think this actually matches web reality in Chrome, does it? The relevant [v8 bug](https://bugs.chromium.org/p/v8/issues/detail?id=896) is still open, and I'm seeing that e.g. `if (false) /a{2,3}/` does not throw an error.

In particular, it looks like in v8 the error is thrown when the literal is *evaluated*, not when it's parsed (but still before it's exec'd).

**Comment by bakkot** [comment]
Needs consensus PR seems right to me.

File: spec.html (hunk 1 above)
Comment (littledan) [reply spec.html]:
```
$ eshost -e "/a{5,3}/"
#### jsc
SyntaxError: Invalid regular expression: numbers out of order in {} quantifier

#### chakracore
SyntaxError: Syntax error in regular expression

#### spidermonkey
SyntaxError: numbers out of order in {} quantifier.:

#### d8
SyntaxError: Invalid regular expression: /a{5,3}/: numbers out of order in {} quantifier
```

File: spec.html (hunk 2 above)
Comment (littledan) [reply spec.html]:
```
$ eshost -e "/[z-a]/"
#### chakracore
SyntaxError: Invalid range in character set

#### jsc
SyntaxError: Invalid regular expression: range out of order in character class

#### spidermonkey
SyntaxError: invalid range in character class:

#### d8
SyntaxError: Invalid regular expression: /[z-a]/: Range out of order in character class
```

File: spec.html (hunk 3 above)
Comment (littledan) [reply spec.html]:
```
$ eshost -e "/{1}/"
#### jsc
SyntaxError: Invalid regular expression: nothing to repeat

#### chakracore
SyntaxError: Unexpected quantifier

#### spidermonkey
SyntaxError: nothing to repeat:

#### d8
SyntaxError: Invalid regular expression: /{1}/: Nothing to repeat
```

File: spec.html (hunk 4 above)
Comment (littledan) [reply spec.html]: ins is used elsewhere in Annex B, e.g., [here](https://tc39.github.io/ecma262/#sec-switch-duplicates-allowed-static-semantics).

```
$ eshost -e "/[\w-\b]/u"
#### jsc
SyntaxError: Invalid regular expression: invalid escaped character for unicode pattern

#### chakracore
/[\w-\b]/u

#### spidermonkey
SyntaxError: character class escape cannot be used in class range in regular expression:

#### d8
SyntaxError: Invalid regular expression: /[\w-\b]/: Invalid character class
```

**Comment by littledan** [comment]
I want to clarify the tests that I included above. It's true that V8 has a known issue where, if a RegExp isn't reached in executing code, it's not parsed and early errors don't come.

However, there are still separate stages of parsing/compiling and executing on a string in V8. You can see this, for example, if you do `/\5/u.toString()`. In the current specification, that would not thrown an error, but in V8, it would. This is the same time as errors for RegExps which don't match the grammar are thrown, for example `/?/`.

I don't think we need to hold back this PR just because V8 has a known spec violation. Whether parsing errors are thrown when parsing the enclosing code, or when evaluating the RegExp for the first time, is not an open spec question given 3/4 of these engines giving early errors:

```
$ eshost -e "false ? /{1}/ : undefined"
#### jsc
SyntaxError: Invalid regular expression: nothing to repeat

#### chakracore
SyntaxError: Unexpected quantifier

#### spidermonkey
SyntaxError: nothing to repeat:

#### d8
undefined
```

It's currently web reality already that, when RegExps are first parsed/compiled, the errors are thrown, rather than later when they are going through a string. That's the delta that this PR makes; it doesn't change whether RegExp syntax errors come at the broader early error time or at a special, later time as V8 does.

**Comment by littledan** [comment]
@bakkot For this to be marked "needs consensus": I think we ended up deciding that not all normative changes need to be discussed in committee, we should just discuss the ones that seem major or don't agree on here. Do you have doubts about this proposal, beyond the fact that V8 doesn't comply with the spec here (in a way that, I argued above, is unrelated)? If not, maybe we can land this patch without spending committee time on it. If you do have doubts, I'll add it to the agenda.

**Comment by allenwb** [comment]
I believe the original assertion that this isn't already covered made in https://github.com/tc39/ecma262/issues/918 is incorrect.  See the comment I left there https://github.com/tc39/ecma262/issues/918#issuecomment-329024576

**Comment by allenwb** [comment]
Below, I'm copying over most of https://github.com/tc39/ecma262/issues/918#issuecomment-329217142

The TL;DR summary is: This patch and all of the other existing syntax error checks in the RegExp evaluation algorithms should be refactored into Static Semantics Early Error rules.

----
The algorithmic syntax error detection in the RegExp evaluation algorithms really shouldn't be there. They are a carry over from Edition 3 where the spec.  which did not use separate static semantic clauses listing syntax-derived early errors (and hence the need for the Ch 16 rule).

What I should have done in ES6 is to factor all of those syntax error checks out of the RegExp evaluation algorithms and restated them as early error rules in [21.2.1.1](https://tc39.github.io/ecma262/#sec-patterns-static-semantics-early-errors). Not doing so was an oversight on my part, or perhaps I ran out of time.

The purpose of placing these errors into 21.2.1.1  (and with the early error static semantic errors, in general) is to clearly state the errors that identify malformed ES code that can be detected and reported independently of an run-time context or values.  The goal is to eliminate implementation variations in when such errors are reported.

**Comment by allenwb** [comment]
Yes, I think it is probably fine.  For some reason, I thought I was seeing something different (new syntax error checks being added to RegExp evaluation rules) when I looked at the patch before sending my previous message

Gawd, I hate reading raw EMU and out of (global) context patch lines.

**Comment by mathiasbynens** [comment]
[This PR gained consensus at the September 2017 TC39 meeting.](https://github.com/rwaldron/tc39-notes/blob/master/es8/2017-09/sep-26.md#12ii-early-errors-for-regexp-literals) Since the Unicode properties proposal is [blocked](https://github.com/tc39/proposal-regexp-unicode-property-escapes/issues/29) on this change, please merge this as soon as possible.

**Comment by bterlson** [comment]
Sorry, I didn't know it was blocking. Happy to merge now. Result of investigation alluded to above: seems great. There are a few differences with what ChakraCore implements but are not concerning and I'll just track them as bugs :)

**Comment by mathiasbynens** [comment]
Thanks!

File: spec.html [hunk 6]
```diff
@@ -30152,12 +30247,12 @@ <h1>NonemptyClassRanges</h1>
           <h1>Runtime Semantics: CharacterRange ( _A_, _B_ )</h1>
           <p>The abstract operation CharacterRange takes two CharSet parameters _A_ and _B_ and performs the following steps:</p>
           <emu-alg>
-            1. If _A_ does not contain exactly one character or _B_ does not contain exactly one character, throw a *SyntaxError* exception.
+            1. Assert: _A_ and _B_ each contain exactly one character.
```
Comment (mathiasbynens) [inline spec.html]: Doesn't e.g. `/[\p{Hex}-\uFFFF]/u` hit this case as well? The assert seems invalid in that case. What am I missing?

Comment (mathiasbynens) [reply spec.html]: It doesn't because of https://github.com/tc39/test262/pull/1014/files/4843f049fafb6cdd61e3be011e3701cc94a9fe46#r200694312. Nothing to see here folks, move along! :)
